Add generic Ore polynomial module

rburing opened the pull request:
This contains only the basic structure so far, such as memory management, additive arithmetic, and multiplication ~~from the left by an element of the base ring~~.

- Part of https://github.com/flintlib/flint/issues/1881

Feedback welcome!

We would like to use this data type as input to differential equation solvers, so we don't need the multiplicative structure initially, though it would be nice to have it in the future.

Multiplicative operations (to be defined later) depend on the Ore context object, so I figured that *all* operations should take the Ore context as input, to make the interface uniform. So now the interface looks like `gr_mpoly`, and the basic operations implemented so far call `gr_poly` methods.

~~Since the base ring will typically be polynomial, should `gr_ore_poly_scalar_mul` and `GR_ORE_POLY_ELEM_CTX` be named differently? Maybe using `base` / `BASE`?~~

~~I also tried to make the module "optional" in the sense that `gr_mpoly` is, let me know if I missed something in this regard.~~

cc @mezzarobba

rburing committed:
Add generic Ore polynomial module

This contains only the basic structure so far, such as memory
management, additive arithmetic, and multiplication from the
left by an element of the base ring.

rburing commented on src/ore_poly_types.h:
```diff
@@ -0,0 +1,30 @@
+extern "C" {
+#endif
+
+typedef enum {
+    ORE_OPERATOR_DERIVATIVE,
+    ORE_OPERATOR_EULER_DERIVATIVE,
+    /* TODO: Shift operator, etc. */
+}
+ore_operator_t;
```
I named this enum (and the value prefix) improperly. What should the name be?

mezzarobba commented:
On Saturday, 12 April 2025 Ricardo Buring wrote:
> > +typedef enum {
> 
> +    ORE_OPERATOR_DERIVATIVE,
> +    ORE_OPERATOR_EULER_DERIVATIVE,
> +    /* TODO: Shift operator, etc. */
> +}
> +ore_operator_t;
> 
> I named this enum (and the value prefix) improperly. What should the
> name be?

Maybe gr_ore_poly_which_operator, for consistency with 
gr_which_structure?

-- 
Marc

rburing commented:
Some/most of the arithmetic/setter methods (`gr_ore_poly_add` for example) are not getting called in the ring tests.

I'm not sure what I did wrong since I did put them in the method table.

fredrik-johansson commented on src/gr.h:
```diff
@@ -693,6 +693,7 @@ typedef enum
     GR_CTX_PSL2Z, GR_CTX_DIRICHLET_GROUP, GR_CTX_PERM,
     GR_CTX_FEXPR,
     GR_CTX_UNKNOWN_DOMAIN,
+    GR_CTX_GR_ORE_POLY,
```
This should go above the groups, with the rings.

fredrik-johansson commented on src/gr.h:
```diff
@@ -1404,6 +1405,12 @@ void gr_ctx_init_gr_mpoly(gr_ctx_t ctx, gr_ctx_t base_ring, slong nvars, const o
 void gr_ctx_init_fmpz_mpoly_q(gr_ctx_t ctx, slong nvars, const ordering_t ord);
 #endif
 
+/* Generic Ore polynomials */
```
I don't think there needs to be a constructor for this ring in ``gr.h``.

It suffices to have ``gr_ore_poly_ctx_init`` in ``gr_ore_poly.h``.

fredrik-johansson commented:
So, working with Ore polynomials over say R[x], one might indeed want to do mixed operations with operands from either R or R[x]. The term "scalar" being ambiguous, instead of adding many extra methods, perhaps just define ``gr_ore_poly_add_other`` etc. that handles all cases?

fredrik-johansson commented:
I don't see the need for ``ore_poly.h`` and ``ore_poly_types.h``.

fredrik-johansson commented:
Most of the one-line trivial methods could go in a single file.

fredrik-johansson commented:
> Multiplicative operations (to be defined later) depend on the Ore context object, so I figured that all operations should take the Ore context as input, to make the interface uniform.

Yes, this is the right way to do it. The way ``gr_poly`` and ``gr_mat`` work directly with the element context is an exception rather than the rule, and at some point we should add a public interface for those types using a poly/mat context as well.

rburing commented:
I reorganized the code to collect the wrapper functions in one file, and applied all the other suggestions above. I removed all the methods with `scalar` in the name and added some `other` methods, indeed this looks much better.

Let me know if this needs any further changes.

fredrik-johansson commented:
Looks good. The only thing missing is documentation.

fredrik-johansson commented on doc/source/gr_ore_poly.rst:
```diff
@@ -0,0 +1,215 @@
+Type compatibility
+-------------------------------------------------------------------------------
+
+The ``gr_ore_poly`` type has the same data layout as the following
```
This paragraph doesn't really make sense since the base ring is always supposed to be something like a polynomial ring and not for example ``fmpz`` (even if that is legal). It would be enough to say that the layout is the same as ``gr_poly``, and that ``gr_poly`` methods therefore can be used for linear and container operations on a ``gr_ore_poly`` provided that one is careful about providing the right context object.

fredrik-johansson commented:
Something that doesn't seem to be well specified in the documentation is the generator name (set_gen_name / string conversion methods). If I'm in ``R[x][D]``, say, shouldn't ``x`` already be named the base ring while the Ore ring generator is ``D``?

fredrik-johansson commented:
In fact, shouldn't the generator x (given as an element of the base ring) be a parameter to the Ore polynomial ring constructor? In general, the base ring could be something multivariate.

rburing commented:
I updated the context and constructor to include an index of a generator of the base ring, and (hopefully) clarified the documentation. I think the restriction to generators of the base ring (rather than elements of the base ring) is reasonable, and e.g. in the univariate case it saves the user from having to define a variable that will probably never be used (e.g. multiplication by the variable will be replaced by a shift). Let me know if this is alright.

mezzarobba commented:
On Tuesday, 29 April 2025 Ricardo Buring wrote:
> @rburing requested your review on: flintlib/flint#2299 Add generic Ore
> polynomial module.

Sure, I can take a look, but maybe not before next week. Thanks for your 
work on this!

rburing commented:
Another question: How should the `q` parameter for q-shifts or q-derivations be handled? The current `gr_ctx_init_gr_ore_poly` and the struct `_gr_ore_poly_ctx_struct` do not refer to it. Can this be left for later? I guess the main question is: do we want to put an extra parameter (again an index into the list of generators or a `gr_srcptr`?) already in `gr_ctx_init_gr_ore_poly`, or add a separate second constructor later (or break source compatibility by adding an extra parameter to `gr_ctx_init_gr_ore_poly` later)?

fredrik-johansson commented:
Either add a ``q``-parameter which is allowed to be ``NULL``, or add a separate constructor ``gr_ctx_init_gr_ore_poly_q_twist`` or whatever later for this case. I don't think it matters much which solution you choose.

mezzarobba commented on src/gr_ore_poly.h:
```diff
@@ -0,0 +1,217 @@
+
+/* Context object */
+
+void gr_ctx_init_gr_ore_poly(gr_ctx_t ctx, gr_ctx_t base_ring, slong base_var, const ore_poly_which_operator_t operator);
```
Maybe avoid `operator` as a variable name as it is a reserved word in c++? Though I don't see how it would be an issue in practice.

mezzarobba commented on src/gr_ore_poly.h:
```diff
@@ -0,0 +1,217 @@
+void gr_ore_poly_clear(gr_ore_poly_t poly, gr_ore_poly_ctx_t ctx);
+
+GR_ORE_POLY_INLINE gr_ptr
+gr_ore_poly_entry_ptr(gr_ore_poly_t poly, slong i, gr_ore_poly_ctx_t ctx)
```
Why `entry` rather than `coeff`?

mezzarobba commented on src/gr_ore_poly/ctx.c:
```diff
@@ -0,0 +1,263 @@
+    {GR_METHOD_CTX_IS_UNIQUE_FACTORIZATION_DOMAIN,  (gr_funcptr) gr_ore_poly_ctx_is_unique_factorization_domain},
+    {GR_METHOD_CTX_IS_FIELD,            (gr_funcptr) gr_generic_ctx_predicate_false},
+    {GR_METHOD_CTX_IS_THREADSAFE,       (gr_funcptr) gr_ore_poly_ctx_is_threadsafe},
+    {GR_METHOD_CTX_SET_GEN_NAME,        (gr_funcptr) _gr_ore_poly_ctx_set_gen_name},
```
Couldn't this use `_gr_gr_poly_ctx_set_gen_name` instead of duplicating it? But I see you are delegating to `gr_poly` methods in other cases, so I must be missing something.

rburing replied:
The function `_gr_gr_poly_ctx_set_gen_name` is only defined in `src/gr/polynomial.c` and not declared in any header. Also the function is tiny and we might want to change the part about the default variable name (e.g. have different defaults for different algebra types), so I think it's fine.

mezzarobba commented on src/gr_ore_poly/ctx.c:
```diff
@@ -0,0 +1,263 @@
+    {GR_METHOD_ZERO,        (gr_funcptr) gr_ore_poly_zero},
+    {GR_METHOD_ONE,         (gr_funcptr) gr_ore_poly_one},
+    {GR_METHOD_NEG_ONE,     (gr_funcptr) gr_ore_poly_neg_one},
+    {GR_METHOD_I,           (gr_funcptr) gr_ore_poly_i},
```
Same question.

mezzarobba commented on src/gr_ore_poly/ring.c:
```diff
@@ -0,0 +1,357 @@
+{
+    gr_stream_t out;
+    gr_stream_init_file(out, stdout);
+    return gr_ore_poly_write(out, poly, "Dx", ctx);
```
I assume you are using a hardcoded variable name for consistency with `gr_poly`, but I assume `gr_poly_print` does that because it only has the coefficient context at hand, whereas here you have the full context.

mezzarobba commented on src/gr_ore_poly/ring.c:
```diff
@@ -0,0 +1,357 @@
+}
+
+int
+gr_ore_poly_set_other(gr_ore_poly_t poly, gr_srcptr x, gr_ctx_t x_ctx, gr_ore_poly_ctx_t ctx)
```
It would be nice to support the base ring's base ring too for rings of the form R[x][D].

mezzarobba commented on src/gr_ore_poly/ring.c:
```diff
@@ -0,0 +1,357 @@
+}
+
+int
+gr_ore_poly_add_other(gr_ore_poly_t res, const gr_ore_poly_t poly, gr_srcptr x, gr_ctx_t x_ctx, gr_ore_poly_ctx_t ctx)
```
Same remark as for `set_other`.

mezzarobba commented on src/gr_ore_poly/ring.c:
```diff
@@ -0,0 +1,357 @@
+*/
+
+#include <stdio.h>
+#include "fmpz.h"
```
Unused includes.

mezzarobba commented on src/gr_ore_poly/ctx.c:
```diff
@@ -0,0 +1,263 @@
+#include "gr_ore_poly.h"
+#include "gr_generic.h"
+
+static const char * default_var = "Dx";
```
I think I would use `D` rather than `Dx` here, as the latter strongly suggests the operator is the derivation wrt x. Or choose the name based on the operator type when the type is known.

mezzarobba commented on src/gr_ore_poly/test/t-set_str.c:
```diff
@@ -0,0 +1,86 @@
+*/
+
+#include "test_helpers.h"
+#include "ulong_extras.h"
```
Unused includes.

mezzarobba commented on src/gr_ore_poly/test/t-set_str.c:
```diff
@@ -0,0 +1,86 @@
+
+/* This is mostly a copy of src/gr_poly/test/t-set_str.c */
+
+TEST_FUNCTION_START(gr_ore_poly_set_str, state)
```
It may be good to test that expressions like `"Dx*x"` either are correctly converted or correctly return `GR_UNABLE`.

mezzarobba commented:
> I think the restriction to generators of the base ring (rather than elements of the base ring) is reasonable

There are definitely cases where one wants to work with derivations like x²·d/dx or operators relative to a generator of the base ring of the base ring, but they can probably be handled by a future “Ore twist as function pointer” mechanism.

mezzarobba commented:
Btw, what do you think about calling (σ,δ) the “(Ore) twist” instead of the “(Ore) operator”? I suppose “twist” would more properly refer to σ alone, but that would avoid some overloading of the word “operator”.

mezzarobba commented on src/gr_ore_poly/ring.c:
```diff
@@ -0,0 +1,357 @@
+}
+
+int
+gr_ore_poly_set_str(gr_ore_poly_t res, const char * s, const char * d, gr_ore_poly_ctx_t ctx)
```
And similarly, here, there is no real need for taking `d` as input.

mezzarobba commented on src/gr_ore_poly/ctx.c:
```diff
@@ -0,0 +1,263 @@
+*/
+    /* todo: we actually want parse using sparse polynomials
+             before converting to the dense representation, to avoid O(n^2) behavior */
+    {GR_METHOD_SET_STR,     (gr_funcptr) gr_generic_set_str_balance_additions},
```
I don't understand why this is not using `gr_ore_poly_set_str`.

mezzarobba commented:
I'm also wondering if just `gr_ore` would be a better prefix than `gr_ore_poly`. OTOH `gr_more` for a potential future multivariate version may be confusing.

mezzarobba commented on src/gr_ore_poly.h:
```diff
@@ -0,0 +1,217 @@
+
+typedef gr_ore_poly_ctx_struct gr_ore_poly_ctx_t[1];
+
+#define GR_ORE_POLY_CTX(ring_ctx) ((_gr_ore_poly_ctx_struct *)((ring_ctx)))
```
document?

rburing replied:
This (and its generalization) is now tracked in https://github.com/flintlib/flint/issues/2318.

rburing commented:
Thanks for the extensive feedback @mezzarobba, I think it looks better now.

What do you think about using the name `ore_algebra_t` for the enum? With values `ORE_ALGEBRA_STANDARD_DERIVATIVE` and `ORE_ALGEBRA_EULER_DERIVATIVE` for now.

@fredrik-johansson Should I keep `entry` in the name of `gr_ore_poly_entry_ptr` for consistency with `gr_poly`, or use `coeff` instead?

fredrik-johansson commented:
I think the ``gr_poly`` method is misnamed and should use ``coeff`` as well.

mezzarobba commented:
> What do you think about using the name `ore_algebra_t` for the enum? With values `ORE_ALGEBRA_STANDARD_DERIVATIVE` and `ORE_ALGEBRA_EULER_DERIVATIVE` for now.

Sounds good too. I think I would call d/dx `DERIVATIVE` instead of `STANDARD_DERIVATIVE`, and you could add `SHIFT` (or maybe `FORWARD_SHIFT`) and `FORWARD_DIFFERENCE` right away.

mezzarobba committed:
add gr_ctx_init_random_{poly,mpoly,series}

mezzarobba committed:
gr_ore_poly: sigma, delta

mezzarobba commented:
I added support for general (σ,δ) and implemented a number of standard (σ,δ) pairs, see https://github.com/mezzarobba/flint/tree/gr_ore_poly. Any feedback welcome. (Ricardo, feel free to merge my branch into this PR if you like.)

vioneers committed:
added gr_ore_poly multiplication, testing and documentation

rburing committed:
gr_ore_poly: fixes after rebase

Mark some functions as static, add ngens to pass ring test, fix
mismatch of sigma_delta_compose function name in header.

rburing committed:
Fix read past end of buffer in flint_fprintf

rburing committed:
gr_ore_poly_ctx_init_mahler: long -> slong

vioneers committed:
resolved pull request comments

mezzarobba committed:
stricter test for Frobenius case of gr_ore_poly_sigma_delta

mezzarobba committed:
size-limit Ore polynomial multiplication

mezzarobba committed:
tweak ore_poly random context generation and test multpliers

mezzarobba committed:
minor improvements to gr_ore_poly docs

mezzarobba commented:
I've made some tweaks so that the tests don't take too long and merged with main in my branch (see https://github.com/mezzarobba/flint/tree/gr_ore_poly).

@fredrik-johansson There are a lot of small things that could be improved, and we may have overlooked some issues, but I think this could already be useful, so feel free to merge it if you want. @vioneers is working on a follow-up PR implementing Euclidean division and gcd.

fredrik-johansson commented:
Great! Better merge this now as it looks functional; any minor polish that's missing can surely come later.
